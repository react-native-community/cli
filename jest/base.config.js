/**
 * @flow
 */
module.exports = {
  rootDir: '../../',
  testEnvironment: 'node',
};
