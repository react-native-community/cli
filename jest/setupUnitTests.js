jest.mock('@react-native-community/cli-tools', () => {
  return {
    ...jest.requireActual('@react-native-community/cli-tools'),
    fetch: jest.fn(),
    logger: {
      success: jest.fn(),
      info: jest.fn(),
      warn: jest.fn(),
      error: jest.fn(),
      debug: jest.fn(),
      log: jest.fn(),
      setVerbose: jest.fn(),
      isVerbose: jest.fn(),
    },
  };
});

jest.setTimeout(20000);
