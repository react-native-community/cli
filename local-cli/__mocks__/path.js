/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 */

'use strict';

const mockPath = {};

function reset(platform) {
  Object.assign(mockPath, jest.requireActual('path')[platform]);
}

mockPath.mock = {reset};

reset('posix');

module.exports = mockPath;
