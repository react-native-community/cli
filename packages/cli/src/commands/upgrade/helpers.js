// @flow
import https from 'https';

export const fetch = (url: string) =>
  new Promise<string>((resolve, reject) => {
    const request = https.get(url, response => {
      if (response.statusCode < 200 || response.statusCode > 299) {
        reject(
          new Error(`Failed to load page, status code: ${response.statusCode}`)
        );
      }
      const body = [];
      response.on('data', chunk => body.push(chunk));
      response.on('end', () => resolve(body.join('')));
    });
    request.on('error', err => reject(err));
  });
