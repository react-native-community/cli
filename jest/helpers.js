// @flow
import fs from 'fs';
import os from 'os';
import path from 'path';
import {createDirectory} from 'jest-util';
import rimraf from 'rimraf';
import execa from 'execa';
import chalk from 'chalk';
import {Writable} from 'readable-stream';

const CLI_PATH = path.resolve(__dirname, '../packages/cli/build/bin.js');

type RunOptions = {
  nodeOptions?: string,
  nodePath?: string,
  timeout?: number, // kill the process after X milliseconds
  expectedFailure?: boolean,
};

export function run(
  dir: string,
  args?: Array<string>,
  options: RunOptions = {
    expectedFailure: false,
  },
) {
  return spawnCli(dir, args, options);
}

// Runs cli until a given output is achieved, then kills it with `SIGTERM`
export async function runUntil(
  dir: string,
  args: Array<string> | void,
  text: string,
  options: RunOptions = {
    expectedFailure: false,
  },
) {
  const spawnPromise = spawnCliAsync(dir, args, {timeout: 30000, ...options});

  spawnPromise.stderr.pipe(
    new Writable({
      write(chunk, _encoding, callback) {
        const output = chunk.toString('utf8');

        if (output.includes(text)) {
          spawnPromise.kill();
        }

        callback();
      },
    }),
  );

  return spawnPromise;
}

export const makeTemplate = (
  str: string,
): ((values?: Array<any>) => string) => (values?: Array<any>) =>
  str.replace(/\$(\d+)/g, (_match, number) => {
    if (!Array.isArray(values)) {
      throw new Error('Array of values must be passed to the template.');
    }
    return values[number - 1];
  });

export const cleanup = (directory: string) => rimraf.sync(directory);

/**
 * Creates a nested directory with files and their contents
 * writeFiles(
 *   '/home/tmp',
 *   {
 *     'package.json': '{}',
 *     'dir/file.js': 'module.exports = "x";',
 *   }
 * );
 */
export const writeFiles = (
  directory: string,
  files: {[filename: string]: string},
) => {
  createDirectory(directory);
  Object.keys(files).forEach(fileOrPath => {
    const dirname = path.dirname(fileOrPath);

    if (dirname !== '/') {
      createDirectory(path.join(directory, dirname));
    }
    fs.writeFileSync(
      path.resolve(directory, ...fileOrPath.split('/')),
      files[fileOrPath],
    );
  });
};

export const copyDir = (src: string, dest: string) => {
  const srcStat = fs.lstatSync(src);
  if (srcStat.isDirectory()) {
    if (!fs.existsSync(dest)) {
      fs.mkdirSync(dest);
    }
    fs.readdirSync(src).map(filePath =>
      copyDir(path.join(src, filePath), path.join(dest, filePath)),
    );
  } else {
    fs.writeFileSync(dest, fs.readFileSync(src));
  }
};

export const getTempDirectory = (name: string) =>
  path.resolve(os.tmpdir(), name);

function spawnCli(dir: string, args?: Array<string>, options: RunOptions = {}) {
  const {spawnArgs, spawnOptions} = getCliArguments({dir, args, options});

  const result = execa.sync(process.execPath, spawnArgs, spawnOptions);

  handleTestCliFailure(options, result, dir, args);

  return result;
}

function spawnCliAsync(
  dir: string,
  args?: Array<string>,
  options: RunOptions = {},
) {
  const {spawnArgs, spawnOptions} = getCliArguments({dir, args, options});

  try {
    return execa(process.execPath, spawnArgs, spawnOptions);
  } catch (result) {
    handleTestCliFailure(options, result, dir, args);
    return result;
  }
}

function getCliArguments({dir, args, options}) {
  const isRelative = !path.isAbsolute(dir);

  if (isRelative) {
    dir = path.resolve(__dirname, dir);
  }

  const env = Object.assign({}, process.env, {FORCE_COLOR: '0'});

  if (options.nodeOptions) {
    env.NODE_OPTIONS = options.nodeOptions;
  }
  if (options.nodePath) {
    env.NODE_PATH = options.nodePath;
  }

  const spawnArgs = [CLI_PATH, ...(args || [])];
  const spawnOptions = {
    cwd: dir,
    env,
    reject: false,
    timeout: options.timeout || 0,
  };
  return {spawnArgs, spawnOptions};
}

function handleTestCliFailure(options, result, dir, args) {
  if (!options.expectedFailure && result.code !== 0) {
    console.log(`Running CLI failed for unexpected reason. Here's more info:

${chalk.bold('dir:')}    ${dir}
${chalk.bold('args:')}   ${(args || []).join(' ')}
${chalk.bold('stderr:')} ${result.stderr}
${chalk.bold('stdout:')} ${result.stdout}
${chalk.bold('code:')}   ${result.code}`);
  }
}
