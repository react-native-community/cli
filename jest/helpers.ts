import fs from 'fs';
import os from 'os';
import path from 'path';
import {createDirectory} from 'jest-util';
// @ts-ignore jsfile
import rimraf from 'rimraf';
import execa from 'execa';
import chalk from 'chalk';
import slash from 'slash';
// @ts-ignore jsfile
import {Writable} from 'readable-stream';

const CLI_PATH = path.resolve(__dirname, '../packages/cli/build/bin.js');

type RunOptions = {
  nodeOptions?: string;
  nodePath?: string;
  timeout?: number; // kill the process after X milliseconds
  expectedFailure?: boolean;
};

/**
 * Helper function to run CLI command in a given folder
 */
export function runCLI(
  dir: string,
  args?: string[],
  options: RunOptions = {
    expectedFailure: false,
  },
) {
  return spawnScript(process.execPath, [CLI_PATH, ...(args || [])], {
    ...options,
    cwd: dir,
  });
}

// Runs cli until a given output is achieved, then kills it with `SIGTERM`
export async function runUntil(
  dir: string,
  args: string[] | undefined,
  text: string,
  options: RunOptions = {
    expectedFailure: false,
  },
) {
  const spawnPromise = spawnScriptAsync(dir, args || [], {
    timeout: 30000,
    cwd: dir,
    ...options,
  });

  spawnPromise.stderr.pipe(
    new Writable({
      write(chunk: any, _encoding: string, callback: () => void) {
        const output = chunk.toString('utf8');

        if (output.includes(text)) {
          spawnPromise.kill();
        }

        callback();
      },
    }),
  );

  return spawnPromise;
}

export const makeTemplate = (
  str: string,
): ((values?: Array<any>) => string) => (values?: Array<any>) =>
  str.replace(/\$(\d+)/g, (_match, number) => {
    if (!Array.isArray(values)) {
      throw new Error('Array of values must be passed to the template.');
    }
    return values[number - 1];
  });

export const cleanup = (directory: string) => rimraf.sync(directory);

/**
 * Creates a nested directory with files and their contents
 * writeFiles(
 *   '/home/tmp',
 *   {
 *     'package.json': '{}',
 *     'dir/file.js': 'module.exports = "x";',
 *   }
 * );
 */
export const writeFiles = (
  directory: string,
  files: {[filename: string]: string},
) => {
  createDirectory(directory);
  Object.keys(files).forEach(fileOrPath => {
    const dirname = path.dirname(fileOrPath);

    if (dirname !== '/') {
      createDirectory(path.join(directory, dirname));
    }
    fs.writeFileSync(
      path.resolve(directory, ...fileOrPath.split('/')),
      files[fileOrPath],
    );
  });
};

export const copyDir = (src: string, dest: string) => {
  const srcStat = fs.lstatSync(src);
  if (srcStat.isDirectory()) {
    if (!fs.existsSync(dest)) {
      fs.mkdirSync(dest);
    }
    fs.readdirSync(src).map(filePath =>
      copyDir(path.join(src, filePath), path.join(dest, filePath)),
    );
  } else {
    fs.writeFileSync(dest, fs.readFileSync(src));
  }
};

export const getTempDirectory = (name: string) =>
  path.resolve(os.tmpdir(), name);

type SpawnOptions = RunOptions & {
  cwd: string;
};

type SpawnFunction<T> = (
  execPath: string,
  args: string[],
  options: SpawnOptions,
) => T;

export const spawnScript: SpawnFunction<execa.ExecaReturns> = (
  execPath,
  args,
  options,
) => {
  const result = execa.sync(execPath, args, getExecaOptions(options));

  handleTestFailure(execPath, options, result, args);

  return result;
};

const spawnScriptAsync: SpawnFunction<execa.ExecaChildProcess> = (
  execPath,
  args,
  options,
) => {
  try {
    return execa(execPath, args, getExecaOptions(options));
  } catch (result) {
    handleTestFailure(execPath, options, result, args);
    return result;
  }
};

function getExecaOptions(options: SpawnOptions) {
  const isRelative = !path.isAbsolute(options.cwd);

  const cwd = isRelative ? path.resolve(__dirname, options.cwd) : options.cwd;

  const env = Object.assign({}, process.env, {FORCE_COLOR: '0'});

  if (options.nodeOptions) {
    env.NODE_OPTIONS = options.nodeOptions;
  }
  if (options.nodePath) {
    env.NODE_PATH = options.nodePath;
  }

  return {
    cwd,
    env,
    reject: false,
    timeout: options.timeout || 0,
  };
}

function handleTestFailure(
  cmd: string,
  options: SpawnOptions,
  result: {[key: string]: any},
  args: string[] | undefined,
) {
  if (!options.expectedFailure && result.code !== 0) {
    console.log(`Running ${cmd} command failed for unexpected reason. Here's more info:
${chalk.bold('cmd:')}     ${cmd}    
${chalk.bold('options:')} ${JSON.stringify(options)}
${chalk.bold('args:')}    ${(args || []).join(' ')}
${chalk.bold('stderr:')}  ${result.stderr}
${chalk.bold('stdout:')}  ${result.stdout}
${chalk.bold('code:')}    ${result.code}`);
  }
}

export function replaceProjectRootInOutput(output: string, testFolder: string) {
  const regex = new RegExp(`(:\\s").*(${slash(testFolder)})`, 'g');
  return slash(output).replace(regex, '$1<<REPLACED_ROOT>>');
}
