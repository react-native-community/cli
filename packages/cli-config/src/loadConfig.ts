import path from 'path';
import {
  UserDependencyConfig,
  ProjectConfig,
  DependencyConfig,
  UserConfig,
  Config,
} from '@react-native-community/cli-types';
import {
  findProjectRoot,
  version,
  resolveNodeModuleDir,
  UnknownProjectError,
} from '@react-native-community/cli-tools';
import findDependencies from './findDependencies';
import resolveReactNativePath from './resolveReactNativePath';
import {
  readConfigFromDisk,
  readDependencyConfigFromDisk,
} from './readConfigFromDisk';
import assign from './assign';
import merge from './merge';

function getDependencyConfig(
  root: string,
  dependencyName: string,
  finalConfig: Config,
  config: UserDependencyConfig,
  userConfig: UserConfig,
  isPlatform: boolean,
): DependencyConfig {
  return merge(
    {
      root,
      name: dependencyName,
      platforms: Object.keys(finalConfig.platforms).reduce(
        (dependency, platform) => {
          const platformConfig = finalConfig.platforms[platform];
          dependency[platform] =
            // Linking platforms is not supported
            isPlatform || !platformConfig
              ? null
              : platformConfig.dependencyConfig(
                  root,
                  config.dependency.platforms[platform],
                );
          return dependency;
        },
        {} as Config['platforms'],
      ),
    },
    userConfig.dependencies[dependencyName] || {},
  ) as DependencyConfig;
}

/**
 * Loads CLI configuration
 */
function loadConfig(projectRoot: string = findProjectRoot()): Config {
  let lazyProject: ProjectConfig;
  const userConfig = readConfigFromDisk(projectRoot);

  const initialConfig: Config = {
    root: projectRoot,
    get reactNativePath() {
      return userConfig.reactNativePath
        ? path.resolve(projectRoot, userConfig.reactNativePath)
        : resolveReactNativePath(projectRoot);
    },
    reactNativeVersion: 'unknown',
    dependencies: userConfig.dependencies,
    commands: userConfig.commands,
    healthChecks: [],
    platforms: userConfig.platforms,
    get project() {
      if (lazyProject) {
        return lazyProject;
      }

      lazyProject = {};
      for (const platform in finalConfig.platforms) {
        const platformConfig = finalConfig.platforms[platform];
        if (platformConfig) {
          lazyProject[platform] = platformConfig.projectConfig(
            projectRoot,
            userConfig.project[platform] || {},
          );
        }
      }

      return lazyProject;
    },
  };

  // Try our best to figure out what version of React Native we're running. This is
  // currently being used to get our deeplinks working, so it's only worried with
  // the major and minor version.
  try {
    let semver = version.current(initialConfig.reactNativePath);
    if (semver) {
      // Retain only these version, since they correspond with our documentation.
      initialConfig.reactNativeVersion = `${semver.major}.${semver.minor}`;
    }
  } catch (e) {
    // If we don't seem to be in a well formed project, give up quietly.
    if (!(e instanceof UnknownProjectError)) {
      throw e;
    }
  }

  const finalConfig = Array.from(
    new Set([
      ...Object.keys(userConfig.dependencies),
      ...findDependencies(projectRoot),
    ]),
  ).reduce((acc: Config, dependencyName) => {
    const localDependencyRoot =
      userConfig.dependencies[dependencyName] &&
      userConfig.dependencies[dependencyName].root;

    try {
      let root =
        localDependencyRoot ||
        resolveNodeModuleDir(projectRoot, dependencyName);
      let config = readDependencyConfigFromDisk(root, dependencyName);

      const isPlatform = Object.keys(config.platforms).length > 0;

      return assign({}, acc, {
        dependencies: assign({}, acc.dependencies, {
          get [dependencyName](): DependencyConfig {
            return getDependencyConfig(
              root,
              dependencyName,
              finalConfig,
              config,
              userConfig,
              isPlatform,
            );
          },
        }),
        commands: [...acc.commands, ...config.commands],
        platforms: {
          ...acc.platforms,
          ...config.platforms,
        },
        healthChecks: [...acc.healthChecks, ...config.healthChecks],
      }) as Config;
    } catch {
      return acc;
    }
  }, initialConfig);

  return finalConfig;
}

export default loadConfig;
