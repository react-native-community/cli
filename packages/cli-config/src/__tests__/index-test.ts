import path from 'path';
import slash from 'slash';
import loadConfig from '..';
import {cleanup, writeFiles, getTempDirectory} from '../../../../jest/helpers';

let DIR = getTempDirectory('config_test');

jest.dontMock('@react-native-community/cli-tools');

const spy = jest.spyOn(global.console, 'warn').mockImplementation(jest.fn());

const iosPath = slash(
  require.resolve('@react-native-community/cli-platform-ios'),
);
const androidPath = slash(
  require.resolve('@react-native-community/cli-platform-android'),
);

const REACT_NATIVE_MOCK = {
  'node_modules/react-native/package.json': '{}',
  'node_modules/react-native/react-native.config.js': `
    const ios = require("${iosPath}");
    const android = require("${androidPath}");
    module.exports = {
      platforms: {
        ios: {
          linkConfig: ios.linkConfig,
          projectConfig: ios.projectConfig,
          dependencyConfig: ios.dependencyConfig,
        },
        android: {
          linkConfig: android.linkConfig,
          projectConfig: android.projectConfig,
          dependencyConfig: android.dependencyConfig,
        },
      },
    };
  `,
};

// Removes string from all key/values within an object
const removeString = (config, str) =>
  JSON.parse(
    JSON.stringify(config, (_key, value) =>
      // In certain cases, `str` (which is a temporary location) will be `/tmp`
      // which is a symlink to `/private/tmp` on OS X. In this case, tests will fail.
      // Following RegExp makes sure we strip the entire path.
      typeof value === 'string'
        ? slash(value.replace(new RegExp(`(.*)${str}`), '<<REPLACED>>'))
        : value,
    ),
  );

beforeEach(async () => {
  await cleanup(DIR);
  jest.resetModules();
  jest.clearAllMocks();
});

afterEach(async () => await cleanup(DIR));

test('should have a valid structure by default', () => {
  DIR = getTempDirectory('config_test_structure');
  writeFiles(DIR, {
    'react-native.config.js': `module.exports = {
      reactNativePath: "."
    }`,
  });
  const config = loadConfig(DIR);
  expect(removeString(config, DIR)).toMatchSnapshot();
});

test('should return dependencies from package.json', () => {
  DIR = getTempDirectory('config_test_deps');
  writeFiles(DIR, {
    ...REACT_NATIVE_MOCK,
    'node_modules/react-native-test/package.json': '{}',
    'node_modules/react-native-test/ReactNativeTest.podspec': '',
    'package.json': `{
      "dependencies": {
        "react-native": "0.0.1",
        "react-native-test": "0.0.1"
      }
    }`,
  });
  const {dependencies} = loadConfig(DIR);
  expect(removeString(dependencies, DIR)).toMatchSnapshot();
});

test('should read a config of a dependency and use it to load other settings', () => {
  DIR = getTempDirectory('config_test_settings');
  writeFiles(DIR, {
    ...REACT_NATIVE_MOCK,
    'node_modules/react-native-test/package.json': '{}',
    'node_modules/react-native-test/ReactNativeTest.podspec': '',
    'node_modules/react-native-test/react-native.config.js': `module.exports = {
      dependency: {
        platforms: {
          ios: {
            scriptPhases: [
              {
                name: "[TEST] Some Configuration",
                path: "./customLocation/custom.sh",
                execution_position: "after_compile",
                input_files: ["input_file"],
                shell_path: "some/shell/path/bash",
                output_files: ["output_file"],
                input_file_lists: ["input_file_1", "input_file_2"],
                output_file_lists: ["output_file_1", "output_file_2"],
                show_env_vars_in_log: false,
                dependency_file: "/path/to/dependency/file"
              }
            ]
          }
        }
      }
    }`,
    'package.json': `{
      "dependencies": {
        "react-native": "0.0.1",
        "react-native-test": "0.0.1"
      }
    }`,
  });
  const {dependencies} = loadConfig(DIR);
  expect(
    removeString(dependencies['react-native-test'], DIR),
  ).toMatchSnapshot();
});

test('should merge project configuration with default values', () => {
  DIR = getTempDirectory('config_test_merge');
  writeFiles(DIR, {
    ...REACT_NATIVE_MOCK,
    'node_modules/react-native-test/package.json': '{}',
    'node_modules/react-native-test/ReactNativeTest.podspec': '',
    'package.json': `{
      "dependencies": {
        "react-native": "0.0.1",
        "react-native-test": "0.0.1"
      }
    }`,
    'react-native.config.js': `module.exports = {
      reactNativePath: ".",
      dependencies: {
        "react-native-test": {
          platforms: {
            ios: {
              scriptPhases: [{name: "abc", path: "./phase.sh"}]
            }
          },
        }
      }
    }`,
  });
  const {dependencies} = loadConfig(DIR);
  expect(removeString(dependencies['react-native-test'], DIR)).toMatchSnapshot(
    'snapshoting `react-native-test` config',
  );
});

test('should load commands from "react-native-foo" and "react-native-bar" packages', () => {
  DIR = getTempDirectory('config_test_packages');
  writeFiles(DIR, {
    'react-native.config.js': 'module.exports = { reactNativePath: "." }',
    'node_modules/react-native-foo/package.json': '{}',
    'node_modules/react-native-foo/react-native.config.js': `module.exports = {
      commands: [
        {
          name: 'foo-command',
          func: () => console.log('foo')
        }
      ]
    }`,
    'node_modules/react-native-bar/package.json': '{}',
    'node_modules/react-native-bar/react-native.config.js': `module.exports = {
      commands: [
        {
          name: 'bar-command',
          func: () => console.log('bar')
        }
      ]
    }`,
    'package.json': `{
      "dependencies": {
        "react-native-foo": "0.0.1",
        "react-native-bar": "0.0.1"
      }
    }`,
  });
  const {commands} = loadConfig(DIR);
  expect(commands).toMatchSnapshot();
});

test('should not skip packages that have invalid configuration (to avoid breaking users)', () => {
  process.env.FORCE_COLOR = '0'; // To disable chalk
  DIR = getTempDirectory('config_test_skip');
  writeFiles(DIR, {
    'node_modules/react-native/package.json': '{}',
    'node_modules/react-native/react-native.config.js': `module.exports = {
      dependency: {
        invalidProperty: 5
      }
    }`,
    'package.json': `{
      "dependencies": {
        "react-native": "0.0.1"
      }
    }`,
  });
  const {dependencies} = loadConfig(DIR);
  expect(removeString(dependencies, DIR)).toMatchSnapshot(
    'dependencies config',
  );
  expect(spy.mock.calls[0][0]).toMatchSnapshot('logged warning');
});

test('does not use restricted "react-native" key to resolve config from package.json', () => {
  DIR = getTempDirectory('config_test_restricted');
  writeFiles(DIR, {
    'node_modules/react-native-netinfo/package.json': `{
      "react-native": "src/index.js"
    }`,
    'react-native.config.js': 'module.exports = { reactNativePath: "." }',
    'package.json': `{
      "dependencies": {
        "react-native-netinfo": "0.0.1"
      }
    }`,
  });
  const {dependencies} = loadConfig(DIR);
  expect(dependencies).toHaveProperty('react-native-netinfo');
  expect(spy).not.toHaveBeenCalled();
});

test('supports dependencies from user configuration with custom root and properties', () => {
  DIR = getTempDirectory('config_test_custom_root');
  const escapePathSeparator = (value: string) =>
    path.sep === '\\' ? value.replace(/(\/|\\)/g, '\\\\') : value;

  writeFiles(DIR, {
    ...REACT_NATIVE_MOCK,
    'native-libs/local-lib/LocalRNLibrary.podspec': '',
    'react-native.config.js': `
const path = require('path');
const root = path.resolve('${escapePathSeparator(
      DIR,
    )}', 'native-libs', 'local-lib');

module.exports = {
  dependencies: {
    'local-lib': {
      root,
      platforms: {
        ios: {
          podspecPath: "custom-path"
        }
      }
    },
  }
}`,
    'package.json': `{
      "dependencies": {
        "react-native": "0.0.1"
      }
    }`,
  });

  const {dependencies} = loadConfig(DIR);
  expect(removeString(dependencies['local-lib'], DIR)).toMatchInlineSnapshot(`
    Object {
      "name": "local-lib",
      "platforms": Object {
        "android": null,
        "ios": Object {
          "configurations": Array [],
          "podspecPath": "custom-path",
          "scriptPhases": Array [],
        },
      },
      "root": "<<REPLACED>>/native-libs/local-lib",
    }
  `);
});

test('should apply build types from dependency config', () => {
  DIR = getTempDirectory('config_test_apply_dependency_config');
  writeFiles(DIR, {
    ...REACT_NATIVE_MOCK,
    'node_modules/react-native-test/package.json': '{}',
    'node_modules/react-native-test/ReactNativeTest.podspec': '',
    'node_modules/react-native-test/react-native.config.js': `module.exports = {
      dependency: {
        platforms: {
          ios: {
            configurations: ["debug"]
          }
        }
      }
    }`,
    'package.json': `{
      "dependencies": {
        "react-native": "0.0.1",
        "react-native-test": "0.0.1"
      }
    }`,
  });
  const {dependencies} = loadConfig(DIR);
  expect(
    removeString(dependencies['react-native-test'], DIR),
  ).toMatchSnapshot();
});

test('supports dependencies from user configuration with custom build type', () => {
  DIR = getTempDirectory('config_test_apply_custom_build_config');
  writeFiles(DIR, {
    ...REACT_NATIVE_MOCK,
    'react-native.config.js': `module.exports = {
  dependencies: {
    'react-native-test': {
      platforms: {
        ios: {
          configurations: ["custom_build_type"]
        }
      }
    },
  }
}`,
    'node_modules/react-native-test/package.json': '{}',
    'node_modules/react-native-test/ReactNativeTest.podspec': '',
    'node_modules/react-native-test/react-native.config.js':
      'module.exports = {}',
    'package.json': `{
      "dependencies": {
        "react-native": "0.0.1",
        "react-native-test": "0.0.1"
      }
    }`,
  });

  const {dependencies} = loadConfig(DIR);
  expect(
    removeString(dependencies['react-native-test'], DIR),
  ).toMatchSnapshot();
});

test('supports disabling dependency for ios platform', () => {
  DIR = getTempDirectory('config_test_disable_dependency_platform');
  writeFiles(DIR, {
    ...REACT_NATIVE_MOCK,
    'node_modules/react-native-test/package.json': '{}',
    'node_modules/react-native-test/ReactNativeTest.podspec': '',
    'node_modules/react-native-test/react-native.config.js': `
      module.exports = {
        dependency: {
          platforms: {
            ios: null
          }
        }
      }
    `,
    'package.json': `{
      "dependencies": {
        "react-native": "0.0.1",
        "react-native-test": "0.0.1"
      }
    }`,
  });

  const {dependencies} = loadConfig(DIR);
  expect(
    removeString(dependencies['react-native-test'], DIR),
  ).toMatchSnapshot();
});

test('should convert project sourceDir relative path to absolute', () => {
  DIR = getTempDirectory('config_test_absolute_project_source_dir');
  const iosProjectDir = './ios2';
  const androidProjectDir = './android2';
  writeFiles(DIR, {
    ...REACT_NATIVE_MOCK,
    'react-native.config.js': `
      module.exports = {
        project: {
          ios: {
            sourceDir: '${iosProjectDir}',
          },
          android: {
            sourceDir: '${androidProjectDir}',
          }
        }
      }
    `,
    'package.json': `{
      "dependencies": {
        "react-native": "0.0.1"
      }
    }`,
    'ios/Podfile': `
      platform :ios, '12.4'
    `,
    'ios2/Podfile': `
      platform :ios, '12.4'
    `,
    'android/AndroidManifest.xml': `
      <manifest xmlns:android="http://schemas.android.com/apk/res/android"
        xmlns:tools="http://schemas.android.com/tools"
        package="com.coinbase.android">
      </manifest>
    `,
    'android2/AndroidManifest.xml': `
      <manifest xmlns:android="http://schemas.android.com/apk/res/android"
        xmlns:tools="http://schemas.android.com/tools"
        package="com.coinbase.android">
      </manifest>
    `,
  });

  const config = loadConfig(DIR);

  expect(config.project.ios?.sourceDir).toBe(path.join(DIR, iosProjectDir));
  expect(config.project.android?.sourceDir).toBe(
    path.join(DIR, androidProjectDir),
  );
});
