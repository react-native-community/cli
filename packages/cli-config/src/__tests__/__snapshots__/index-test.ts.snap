// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should apply build types from dependency config 1`] = `
Object {
  "name": "react-native-test",
  "platforms": Object {
    "android": null,
    "ios": Object {
      "configurations": Array [
        "debug",
      ],
      "podspecPath": "<<REPLACED>>/node_modules/react-native-test/ReactNativeTest.podspec",
      "scriptPhases": Array [],
    },
  },
  "root": "<<REPLACED>>/node_modules/react-native-test",
}
`;

exports[`should have a valid structure by default 1`] = `
Object {
  "commands": Array [],
  "dependencies": Object {},
  "healthChecks": Array [],
  "platforms": Object {},
  "project": Object {},
  "reactNativePath": "<<REPLACED>>",
  "reactNativeVersion": "unknown",
  "root": "<<REPLACED>>",
}
`;

exports[`should load commands from "react-native-foo" and "react-native-bar" packages 1`] = `
Array [
  Object {
    "func": [Function],
    "name": "foo-command",
  },
  Object {
    "func": [Function],
    "name": "bar-command",
  },
]
`;

exports[`should merge project configuration with default values: snapshoting \`react-native-test\` config 1`] = `
Object {
  "name": "react-native-test",
  "platforms": Object {
    "android": null,
    "ios": Object {
      "configurations": Array [],
      "podspecPath": "<<REPLACED>>/node_modules/react-native-test/ReactNativeTest.podspec",
      "scriptPhases": Array [
        Object {
          "name": "abc",
          "path": "./phase.sh",
        },
      ],
    },
  },
  "root": "<<REPLACED>>/node_modules/react-native-test",
}
`;

exports[`should not skip packages that have invalid configuration (to avoid breaking users): dependencies config 1`] = `
Object {
  "react-native": Object {
    "name": "react-native",
    "platforms": Object {},
    "root": "<<REPLACED>>/node_modules/react-native",
  },
}
`;

exports[`should not skip packages that have invalid configuration (to avoid breaking users): logged warning 1`] = `"warn Package react-native contains invalid configuration: \\"dependency.invalidProperty\\" is not allowed. Please verify it's properly linked using \\"react-native config\\" command and contact the package maintainers about this."`;

exports[`should read a config of a dependency and use it to load other settings 1`] = `
Object {
  "name": "react-native-test",
  "platforms": Object {
    "android": null,
    "ios": Object {
      "configurations": Array [],
      "podspecPath": "<<REPLACED>>/node_modules/react-native-test/ReactNativeTest.podspec",
      "scriptPhases": Array [
        Object {
          "dependency_file": "/path/to/dependency/file",
          "execution_position": "after_compile",
          "input_file_lists": Array [
            "input_file_1",
            "input_file_2",
          ],
          "input_files": Array [
            "input_file",
          ],
          "name": "[TEST] Some Configuration",
          "output_file_lists": Array [
            "output_file_1",
            "output_file_2",
          ],
          "output_files": Array [
            "output_file",
          ],
          "path": "./customLocation/custom.sh",
          "shell_path": "some/shell/path/bash",
          "show_env_vars_in_log": false,
        },
      ],
    },
  },
  "root": "<<REPLACED>>/node_modules/react-native-test",
}
`;

exports[`should return dependencies from package.json 1`] = `
Object {
  "react-native": Object {
    "name": "react-native",
    "platforms": Object {
      "android": null,
      "ios": null,
    },
    "root": "<<REPLACED>>/node_modules/react-native",
  },
  "react-native-test": Object {
    "name": "react-native-test",
    "platforms": Object {
      "android": null,
      "ios": Object {
        "configurations": Array [],
        "podspecPath": "<<REPLACED>>/node_modules/react-native-test/ReactNativeTest.podspec",
        "scriptPhases": Array [],
      },
    },
    "root": "<<REPLACED>>/node_modules/react-native-test",
  },
}
`;

exports[`supports dependencies from user configuration with custom build type 1`] = `
Object {
  "name": "react-native-test",
  "platforms": Object {
    "android": null,
    "ios": Object {
      "configurations": Array [
        "custom_build_type",
      ],
      "podspecPath": "<<REPLACED>>/node_modules/react-native-test/ReactNativeTest.podspec",
      "scriptPhases": Array [],
    },
  },
  "root": "<<REPLACED>>/node_modules/react-native-test",
}
`;

exports[`supports disabling dependency for ios platform 1`] = `
Object {
  "name": "react-native-test",
  "platforms": Object {
    "android": null,
    "ios": null,
  },
  "root": "<<REPLACED>>/node_modules/react-native-test",
}
`;
